File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
const dr = [-1, 1, 0, 0];
const dc = [0, 0, -1, 1];

function solution(maps) {
    return bfs(maps);
}

function bfs(maps){
    const queue = [];
    queue.push([0, 0, 1]);
    const visited = Array.from({length: maps.length}, ()=>new Array(maps[0].length).fill(false));
    visited[0][0]=true;
    while (queue.length>0){
        const [r, c, dist] = queue.shift();
        if (r===maps.length-1 && c===maps[0].length-1){
            return dist;
        }
        for (let d=0; d<4; d++){
            const nr = r+dr[d];
            const nc = c+dc[d];
            if (nr>=0 && nr<maps.length && nc>=0 && nc<maps[0].length && !visited[nr][nc] && maps[nr][nc]===1){
                visited[nr][nc]=true;
                queue.push([nr, nc, dist+1]);
            }
        }
    }
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};

    public int solution(int[][] maps) {
        int answer = 0;
        
        return bfs(maps);
    }
    
    public int bfs(int[][] maps) {
        Queue<Integer[]> queue = new LinkedList<>();
        boolean[][] visited = new boolean[maps.length][maps[0].length];
        queue.offer(new Integer[]{0, 0});
        visited[0][0] = true;
        int cnt = 0;
        while(!queue.isEmpty()){
            int len = queue.size();
            cnt++;
            for (int i=0; i<len; i++){
                Integer[] top = queue.poll();
                for (int d=0; d<4; d++){
                    int nr = top[0]+dr[d];
                    int nc = top[1]+dc[d];
                    if (nr>=0 && nr<maps.length && nc>=0 && nc<maps[0].length && !visited[nr][nc] && maps[nr][nc]==1){
                        if (nr==maps.length-1 && nc==maps[0].length-1) {
                            return cnt+1;
                        }
                        queue.offer(new Integer[]{nr, nc});
                        visited[nr][nc]=true;
                    }
                }
                
            }
        }
        return -1;
    }
}